platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/isolation/IsolatedActionSerializer.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.configurationcache.isolation
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap
+import org.gradle.api.IsolatedAction
+import org.gradle.api.services.internal.BuildServiceProvider
+import org.gradle.configurationcache.extensions.uncheckedCast
+import org.gradle.configurationcache.extensions.useToRun
+import org.gradle.configurationcache.logger
+import org.gradle.configurationcache.problems.ProblemsListener
+import org.gradle.configurationcache.problems.PropertyProblem
+import org.gradle.configurationcache.serialization.ClassDecoder
+import org.gradle.configurationcache.serialization.ClassEncoder
+import org.gradle.configurationcache.serialization.DefaultReadContext
+import org.gradle.configurationcache.serialization.DefaultWriteContext
+import org.gradle.configurationcache.serialization.IsolateOwner
+import org.gradle.configurationcache.serialization.ReadContext
+import org.gradle.configurationcache.serialization.WriteContext
+import org.gradle.configurationcache.serialization.beans.BeanStateReaderLookup
+import org.gradle.configurationcache.serialization.beans.BeanStateWriterLookup
+import org.gradle.configurationcache.serialization.readNonNull
+import org.gradle.configurationcache.serialization.runReadOperation
+import org.gradle.configurationcache.serialization.runWriteOperation
+import org.gradle.configurationcache.serialization.withIsolate
+import org.gradle.configurationcache.services.IsolatedActionCodecsFactory
+import org.gradle.internal.serialize.kryo.KryoBackedDecoder
+import org.gradle.internal.serialize.kryo.KryoBackedEncoder
+import java.io.ByteArrayOutputStream
+import java.io.OutputStream
+import java.util.IdentityHashMap
+
+
+/**
+ * Serialized state of an object graph containing one or more [IsolatedAction]s.
+ */
+class SerializedAction(
+    /**
+     * The serialized graph.
+     */
+    val graph: ByteArray,
+
+    /**
+     * External references that are not serialized directly as part of the [graph].
+     * These include [Class] and [BuildServiceProvider] references.
+     */
+    val environment: Map<Int, Any>,
```

Comment:
🤔 Maybe it worth to document what does the key mean in this map?

Reply:
Done.